pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -213,57 +293,104 @@ func (ap *ApplicationProfileCacheImpl) removeContainer(containerID string) {
-
-	apState := newApplicationProfileState(appProfile)
-	ap.slugToState.Set(apName, apState)
-
-	// the cache holds only completed application profiles.
-	// check if the application profile is completed
-	// if status was completed and now is not (e.g. mode changed from complete to partial), remove from cache
-	if apState.status != helpersv1.Completed {
-		if ap.slugToAppProfile.Has(apName) {
-			ap.slugToAppProfile.Delete(apName)
-			ap.allProfiles.Remove(apName)
-		}
-		return
-	}
-
-	// add to the cache
-	ap.allProfiles.Add(apName)
-
-	if ap.slugToContainers.Has(apName) {
-		// get the full application profile from the storage
-		// the watch only returns the metadata
-		// avoid thundering herd problem by adding a random delay
-		time.AfterFunc(utils.RandomDuration(ap.maxDelaySeconds, time.Second), func() {
-			ap.addFullApplicationProfile(appProfile, apName)
-		})
-	}
-}
 
 func (ap *ApplicationProfileCacheImpl) addFullApplicationProfile(appProfile *v1beta1.ApplicationProfile, apName string) {
 	fullAP, err := ap.getApplicationProfile(appProfile.GetNamespace(), appProfile.GetName())
 	if err != nil {
 		logger.L().Error("failed to get full application profile", helpers.Error(err))
 		return
 	}
+
+	// Check if there's a pending user-managed profile to merge
+	if ap.userManagedProfiles.Has(apName) {
+		userManagedProfile := ap.userManagedProfiles.Get(apName)
+		fullAP = ap.performMerge(fullAP, userManagedProfile)
+		// Clean up the user-managed profile after successful merge
+		ap.userManagedProfiles.Delete(apName)
+		logger.L().Debug("merged pending user-managed profile", helpers.String("name", apName))
+	}
+
 	ap.slugToAppProfile.Set(apName, fullAP)
 	for _, i := range ap.slugToContainers.Get(apName).ToSlice() {
 		ap.containerToSlug.Set(i, apName)
 	}
 	logger.L().Debug("added pod to application profile cache", helpers.String("name", apName))
 }
 
+func (ap *ApplicationProfileCacheImpl) performMerge(normalProfile, userManagedProfile *v1beta1.ApplicationProfile) *v1beta1.ApplicationProfile {
+	mergedProfile := normalProfile.DeepCopy()
+
+	// Merge spec
+	mergedProfile.Spec.Containers = ap.mergeContainers(mergedProfile.Spec.Containers, userManagedProfile.Spec.Containers)
+	mergedProfile.Spec.InitContainers = ap.mergeContainers(mergedProfile.Spec.InitContainers, userManagedProfile.Spec.InitContainers)
+	mergedProfile.Spec.EphemeralContainers = ap.mergeContainers(mergedProfile.Spec.EphemeralContainers, userManagedProfile.Spec.EphemeralContainers)
+
+	// Remove the user-managed annotation
+	delete(mergedProfile.Annotations, "kubescape.io/managed-by")
+
+	return mergedProfile
+}
+
+func (ap *ApplicationProfileCacheImpl) mergeContainers(normalContainers, userManagedContainers []v1beta1.ApplicationProfileContainer) []v1beta1.ApplicationProfileContainer {
+	// Create a map to store containers by name
+	containerMap := make(map[string]int) // map name to index in slice
+
+	// Store indices of normal containers
+	for i := range normalContainers {
```

Comment:
we don't need that... the order of containers should be kept as in the Pod for both profiles (otherwise I will complain to @jnathangreeg )
maybe something strange could appear if we have ephemeral containers added in either profile (always appended at the right place, leaving empty slices if needed)

Reply:
done